### PR TITLE
Add support for pubnub sources

### PR DESCRIPTION
**Pull request by petrosagg**
This PR adds basic support for pubnub sources and also acts as a reference implementation of a source using the recently merged SimpleSource trait.

The syntax for the source creation is: 

```sql
CREATE SOURCE foo FROM PUBNUB SUBSCRIBE KEY 'sub_key' CHANNEL 'channel'
```

This sources allows users to test out materialize using one of the example realtime streams offered by pubnub. The streams are listed here https://www.pubnub.com/developers/realtime-data-streams/

**Caveat**: The wikipedia stream was a bit unreliable in my testing but the other ones work great

Here is an example usage of the market order stream:

```sql
# The raw source that brings in data as a single text column containing JSON
CREATE SOURCE market_orders_raw
FROM PUBNUB
    SUBSCRIBE KEY 'sub-c-4377ab04-f100-11e3-bffd-02ee2ddab7fe'
    CHANNEL 'pubnub-market-orders';

# Extract a couple JSON fields
CREATE VIEW market_orders AS
    SELECT
        val->>'symbol' AS symbol,
        (val->'bid_price')::float AS bid_price
FROM (SELECT text::jsonb AS val FROM market_orders_raw);

# Create a materialized aggregation
CREATE MATERIALIZED VIEW avg_bid AS
    SELECT symbol, AVG(bid_price) FROM market_orders GROUP BY symbol;

# See the data change
SELECT * FROM avg_bid;
```

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6189)

### CODE REVIEW

File: src/dataflow/src/source/pubnub.rs
Comment (mjibson) [inline src/dataflow/src/source/pubnub.rs:39]: How does this handle connection failures mid stream? Can it resume from where it left off?

Comment (petrosagg) [reply src/dataflow/src/source/pubnub.rs:39]: It doesn't, I can try to add reconnect logic bug messages may be lost. That's just the nature of pubnub though so it's ok

**Review by benesch — COMMENTED** [review]
😍 

Just two minor comments:

  * s/Pubnub/PubNub/ throughout
  * Do we want to mark as experimental for a release or two before we commit to supporting this?

**Comment by petrosagg** [comment]
@benesch I thought I marked experimental with this line https://github.com/MaterializeInc/materialize/pull/6189/files#diff-c840a7fbb6a642d6cbf71dddd92ec1b6dfa86310022ead5673e8fa509ba73dc4R753

Is there some other place that I need to change?

Names fixed!

**Review by cirego — APPROVED** [review]
Awesome! Simple source looks really easy to use!

This will be really nice for people trying out the cloud product 🥳

File: src/dataflow/Cargo.toml
Comment (benesch) [inline src/dataflow/Cargo.toml:49]: Aw, shoot, too bad about having to fork. In the interest of reproducible builds in the future, can you create a MaterializeInc-owned fork, and merge your Tokio v1.0 upgrade into `master`? Feature branches on personal forks have a way of disappearing, and then it's impossible to build that version of Materialize in the future.

**Review by benesch — APPROVED** [review]
LGTM, modulo one small logistical issue with feature branches on personal forks!

File: src/dataflow/Cargo.toml
Comment (petrosagg) [reply src/dataflow/Cargo.toml:49]: My plan was to put the PR out for review in a runnable state (hence the personal fork) and merge once my upstream PR https://github.com/pubnub/rust/pull/61 is merged and released.

I'll give it a few days and implement your proposal if we're still blocked

Comment (benesch) [reply src/dataflow/Cargo.toml:49]: You are more optimistic than I! Most crates are on a release cadence of several months, if not years. :) Sounds good, though.

**Comment by philip-stoev** [comment]
@petrosagg  Can we add some simple test for that? I can give it a try if you keep the review open for another day or two.

**Comment by petrosagg** [comment]
@philip-stoev I was thinking about that, how would you go about it since it's using an external service?

**Comment by petrosagg** [comment]
using the example streams the test can go as far as creating the source and verifying that a datapoint came through

**Comment by philip-stoev** [comment]
@petrosagg yes, some example from their documentation would be Ok. I would also add some tests around wrong URLs, wrong security tokens and the like.

You are right that it is not ideal that we are accessing an external service, but we seem to do that quite a lot already in our CI , that is, even the linter downloads half of crates.io, not to mention the stuff we download from the docker hub. If this lives in a separate mzcompose file, then a failure of this test would not bring down any others.

File: test/pubnub/mzcompose.yml
Comment (philip-stoev) [inline test/pubnub/mzcompose.yml:1]: You need to register this file in ```ci/test/pipeline.template.yml``` so that it gets run by the CI.

Comment (petrosagg) [reply test/pubnub/mzcompose.yml:1]: done

File: src/dataflow/Cargo.toml
Comment (petrosagg) [reply src/dataflow/Cargo.toml]: I'm too excited to wait. Created the mz-owned fork and will merge as soon as tests pass!

**Comment by benesch** [comment]
Can always fix up in post but I am uncomfortable with the idea of adding a dependency on PubNub to our CI. While we do depend on GitHub, Docker Hub, crates.io, and NPM, those are a bit more fundamental, and even still I’d love to get us to a place where the only external service CI depends on is GitHub and a Docker registry we run ourselves.

We have some tests that run against Kinesis and S3 in CI presently, and they’re brutally flaky. We ought to make those use local stack only in PRs, and then set up some nightly stuff that runs those tests against the real services. And then ideally do that for PubNub too.

I’m driving so can’t say much more now. Definitely don’t feel like this needs to be addressed before you merged; just wanted to jump in to say that the existing external dependencies in CI should not be used to justify adding new external dependencies, because they are already problematic.

**Comment by petrosagg** [comment]
That's a fair point, I'm going to remove it from the pipeline until we get some time to talk through this then.

My main motivation behind adding the dependency was not so much that we already have external dependencies in our CI but that this will potentially become the first experience anyone has with the cloud product and so it must always work.

**Pull request merged**
